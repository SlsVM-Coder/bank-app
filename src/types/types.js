export const types = {
  login: '[Auth] Login',
  logout: '[Auth] Logout',

  uiSetError: '[UI] Set error',
  uiRemoveError: '[UI] Remove error',
  uiStartLoading: '[UI] Start loading',
  uiFinishLoading: '[UI] Finish loading',

  dataLoaded: '[Money] Load data',
  addMovement: '[Money] Add movement',
  updateBalance: '[Money] Update balance',
  moneyLogout: '[Money] Money logout cleaning',
};
