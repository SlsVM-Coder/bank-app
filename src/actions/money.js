import { types } from '../types/types.js';
import { db } from '../firebase/firebaseConfig';
import { loadData } from './../helpers/loadData';

export const startLoadingData = (email) => {
  return async (dispatch) => {
    const data = await loadData(email);
    dispatch(dataLoaded(data));
  };
};

export const dataLoaded = (data) => ({
  type: types.dataLoaded,
  payload: data,
});

export const depositMoney = (quantity) => {
  return async (dispatch, getState) => {
    const { email, name } = getState().auth;
    const { balance } = getState().money;
    const newQuantity = balance + quantity;
    dispatch(startUpdateBalance(email, newQuantity));
    const movement = {
      sender: name,
      receiver: name,
      quantity,
      action: 'Deposit',
      date: Date.now(),
    };
    dispatch(startAddMovement(email, movement));
  };
};

export const transferMoney = (quantity, receiverEmail) => {
  return async (dispatch, getState) => {
    debugger;
    const { email, name: senderName } = getState().auth;
    const { balance } = getState().money;
    const snapReceiverInfo = await db.doc(`${receiverEmail}/info`).get();
    const { name: receiverName } = snapReceiverInfo.data();
    const newQuantity = balance - quantity;
    dispatch(startUpdateBalance(email, newQuantity)); // users update

    const movement = {
      sender: senderName,
      receiver: receiverName,
      quantity,
      action: 'Transfer',
      date: Date.now(),
    };
    dispatch(startAddMovement(email, movement));

    const snapReceiverBalance = await db.collection(`${receiverEmail}/`).get();
    const [{ balance: receiverBalance }] = snapReceiverBalance.docs.map((doc) =>
      doc.data(),
    );
    const newQuantityReceiver = receiverBalance + quantity;
    dispatch(startUpdateBalance(receiverEmail, newQuantityReceiver)); // receiver update

    const receiverMovement = {
      sender: senderName,
      receiver: receiverName,
      quantity,
      action: 'Deposit',
      date: Date.now(),
    };
    dispatch(startAddMovement(receiverEmail, receiverMovement));
  };
};

export const startUpdateBalance = (email, newQuantity) => {
  return async (dispatch, getState) => {
    await db.doc(`/${email}/data/`).update({ balance: newQuantity });

    const { email: userEmail } = getState().auth;

    if (userEmail === email) {
      dispatch(updateBalance(newQuantity));
    }
  };
};

export const updateBalance = (quantity) => ({
  type: types.updateBalance,
  payload: quantity,
});

export const startAddMovement = (email, movement) => {
  return async (dispatch) => {
    const doc = await db.collection(`${email}/data/movements`).add(movement);
    dispatch(addMovement(doc.id, movement));
  };
};

export const addMovement = (id, movement) => ({
  type: types.addMovement,
  payload: {
    id,
    ...movement,
  },
});

export const moneyLogout = () => ({
  type: types.moneyLogout,
});
