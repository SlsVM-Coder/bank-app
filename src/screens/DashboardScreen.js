import React from 'react';
import { useSelector } from 'react-redux';
import { BalanceCard } from './../components/BalanceCard';

export const DashboardScreen = () => {
  const { name } = useSelector((state) => state.auth);

  return (
    <div>
      <h1>Welcome back, {name}</h1>
      <p>You can:</p>
      <p>Check your money, transfer money or view your last movements. </p>
      <BalanceCard />
    </div>
  );
};
