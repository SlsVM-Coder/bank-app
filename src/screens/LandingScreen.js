import React from 'react';
import { Link } from 'react-router-dom';
import { Layout } from '../components/Layout';

export const LandingScreen = () => {
  return (
    <Layout
      component="Landing"
      title="Your favourite online wallet"
      img="landingImg"
    >
      <Link to="/auth">
        <button renderas="button" className="btn-primary">
          <span>Get started</span>
        </button>
      </Link>
    </Layout>
  );
};
