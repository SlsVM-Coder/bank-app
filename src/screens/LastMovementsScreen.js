import React from 'react';
import { useSelector } from 'react-redux';
import { Movement } from './../components/Movement';
import { Layout } from './../components/Layout';

export const LastMovementsScreen = () => {
  const { movements } = useSelector((state) => state.money);

  return (
    <Layout
      component="LastMovements"
      img="actionsImg"
      title="Last movements"
    >
      <div className="MovementsContainer">
        {movements.length > 0 ? (
          movements.map((movement, index) => (
            <Movement
              key={`${index}-${movement.date}`}
              {...movement}
            />
          ))
        ) : (
          <p>You don't have any movements yet.</p>
        )}
      </div>
    </Layout>
  );
};
