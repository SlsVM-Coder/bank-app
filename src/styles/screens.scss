.Dashboard {
  .ButtonsContainer {
    display: flex;
    flex-direction: column;
    margin-top: 20px;

    button {
      width: 200px;
      margin-bottom: 5px;
    }
  }
}

.LastMovements {
  .MovementsContainer {
    max-height: 800px;
    overflow: auto;
  }
}

.otherAuth {
  margin-top: 10px;
  display: flex;
  align-items: center;

  .link {
    margin-left: 10px;
  }
}

.msgError {
  width: 250px;
  background-color: #ab0d0d;
  color: white;
  padding: 10px 0;
  border-radius: $border-radius;
  padding-left: 10px;
  margin: 10px 0 20px 0;
}
