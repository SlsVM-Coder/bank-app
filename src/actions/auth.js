import Swal from 'sweetalert2';
import { types } from '../types/types.js';
import { db, firebase } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import { startLoadingData, moneyLogout } from './money';

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(startLoadingData(user.email));
        dispatch(login(user.uid, user.email, user.displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        dispatch(finishLoading());
        Swal.fire('Error', e.message, 'error');
      });
  };
};
export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    return firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        const info = {
          uid: user.uid,
          email: user.email,
          name: user.displayName,
        };
        await db.doc(`${user.email}/info`).set(info);
        await db.doc(`/${user.email}/data/`).set({ balance: 0 });
        dispatch(login(user.uid, user.email, user.displayName));
        dispatch(startLoadingData(user.email));
      })
      .catch((e) => {
        console.log(e);
        Swal.fire('Error', e.message, 'error');
      });
  };
};

export const login = (uid, email, displayName) => ({
  type: types.login,
  payload: {
    uid,
    email,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(moneyLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
