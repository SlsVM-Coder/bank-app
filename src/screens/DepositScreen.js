import React from 'react';
import { useForm } from './../hooks/useForm';
import { depositMoney } from './../actions/money';
import { useDispatch } from 'react-redux';

export const DepositScreen = () => {
  const dispatch = useDispatch();
  const [formValues, handleInputChange] = useForm({
    quantity: '',
  });
  const { quantity } = formValues;

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(depositMoney(parseFloat(quantity)));
  };

  return (
    <div>
      <h1>deposit</h1>
      <form onSubmit={handleSubmit}>
        <label htmlFor="quantity">Quantity:</label>
        <input
          type="number"
          placeholder="0"
          name="quantity"
          id="quantity"
          className="input-text"
          value={quantity}
          onChange={handleInputChange}
        />
      </form>
    </div>
  );
};
