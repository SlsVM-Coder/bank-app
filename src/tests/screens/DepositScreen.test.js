import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';

import { types } from './../../types/types';
import { DepositScreen } from './../../screens/DepositScreen';
import { depositMoney } from './../../actions/money';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    uid: 'test uid',
    name: 'Name Lastname',
    email: 'test-bank@testing.com',
  },
  ui: {
    loading: false,
    msgError: null,
  },
  money: {
    balance: 50,
    movements: [],
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <DepositScreen />
  </Provider>,
);
jest.mock('../../actions/money', () => ({
  depositMoney: jest.fn(),
}));

describe('Pruebas en <DepositScreen />', () => {
  beforeEach(() => {
    store = mockStore(initState);
    jest.clearAllMocks();
  });

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de hacer el dispatch de la acciÃ³n respectiva', () => {
    wrapper.find('form').prop('onSubmit')({
      preventDefault() {},
    });
    expect(depositMoney).toHaveBeenCalled();
  });
});
