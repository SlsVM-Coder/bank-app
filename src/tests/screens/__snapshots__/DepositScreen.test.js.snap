// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DepositScreen /> debe de mostrarse correctamente 1`] = `
<div
  className="Container Deposit"
>
  <div
    className="LeftContent"
  >
    <div
      className="Content"
    >
      <h1>
         
        Deposit money
         
      </h1>
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="quantity"
        >
          Quantity:
        </label>
        <input
          className="input-text"
          id="quantity"
          min="0"
          name="quantity"
          onChange={[Function]}
          placeholder="0â‚¬"
          type="number"
          value=""
        />
        <button
          className="btn-primary"
        >
          Deposit
        </button>
      </form>
    </div>
  </div>
  <div
    className="RightContent"
  >
    <img
      alt="actionsImg"
      src="actionsImg.png"
    />
  </div>
</div>
`;
